src/powershell/Program.cs

```diff
@@ -19,7 +59,439 @@ public sealed class ManagedPSEntry
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
+            {
+                return false;
+            }
+
+            // Check arg chars in order and allow prefixes
+            for (int i = 1; i < arg.Length; i++)
+            {
+                if (arg[i] != paramToCheck[i-1]
+                    && arg[i] != paramToCheckUpper[i-1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Create the exec call to /bin/{z}sh -l -c 'exec -a +pwsh pwsh "$@"' and run it.
+        /// </summary>
+        /// <param name="args">The argument vector passed to pwsh.</param>
+        /// <param name="loginArgIndex">The index of -Login in the argument vector.</param>
+        /// <param name="isMacOS">True if we are running on macOS.</param>
+        /// <param name="pwshPath">Absolute path to the pwsh executable.</param>
+        /// <returns>
+        /// The exit code of exec if it fails.
+        /// If exec succeeds, this process is overwritten so we never actually return.
+        /// </returns>
+        private static int ExecPwshLogin(string[] args, int loginArgIndex, string pwshPath, bool isMacOS)
+        {
+            // Create input for /bin/sh that execs pwsh
+            int quotedPwshPathLength = GetQuotedPathLength(pwshPath);
+
+            // /bin/sh does not support the exec -a feature
+            int pwshExecInvocationLength = isMacOS
+                ? quotedPwshPathLength + 19  // exec -a +pwsh '{pwshPath}' "$@"
+                : quotedPwshPathLength + 10; // exec '{pwshPath}' "$@"
+
+            string pwshInvocation = string.Create(
+                pwshExecInvocationLength, 
+                (pwshPath, quotedPwshPathLength, isMacOS),
+                CreatePwshInvocation);
+
+            // Set up the arguments for /bin/sh
+            var execArgs = new string[args.Length + 5];
+
+            // execArgs[0] is set below to the correct shell executable
+
+            // The command arguments
+
+            // First argument is the command name.
+            // Setting this to /bin/sh enables sh emulation in zsh (which examines $0 to determine how it should behave).
+            execArgs[0] = "/bin/sh"; 
```

Comment:
We run `zsh` on macOS, should this arg be `/bin/zsh` then?

Reply:
Using `/bin/sh` means `zsh` starts in `sh` emulation mode. I might try enhancing the comment above to make that clearer